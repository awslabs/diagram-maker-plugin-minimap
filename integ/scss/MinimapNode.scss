.example-minimap-node {
    box-sizing: border-box;
    border: 1px solid #545b64;
    background: #ffffff;
    padding: 0px;
    height: 100%;
    transition: all .2s ease;

    &.odd {
      border: 1px solid red;
    }

    &.even {
      border: 1px solid blue;
    }

    &.selected {
      border: 1px solid #0073bb;;
    }

    &.rectangle {
      padding: 1px;
    }

    &.circle {
      border-radius: 100%;
      padding: 0px;
      position: absolute;
      width: 100%;
      height: 100%;
      top: 0;
      left: 0;
    }
}
