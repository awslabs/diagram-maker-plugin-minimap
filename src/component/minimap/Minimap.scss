.dm-minimap-canvas {
  position: absolute;
  background-color: #f2f3f3;
}
.dm-minimap-rectangle{
  border: 1px solid #ff9900;
  position: absolute;
  z-index: 4000;
}
.dm-minimap-container{
  position: relative;
  background-color: #d5dbdb;
}
