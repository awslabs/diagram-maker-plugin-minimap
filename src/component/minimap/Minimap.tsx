import {
  DestroyCallback,
  DiagramMakerData,
  DiagramMakerNode,
  DiagramMakerNodes,
  DiagramMakerWorkspace,
  Size
} from 'diagram-maker';
import MiniNode from 'diagramMakerMinimap/component/mininode/MiniNode';
import {
  DiagramMakerMinimapType,
  getRectOffset,
  getRectSize,
  getScale,
  RenderCallback
} from 'diagramMakerMinimap/service';
import * as React from 'react';
import './Minimap.scss';

export interface MinimapProps<NodeType, EdgeType> {
  state: DiagramMakerData<NodeType, EdgeType>;
  renderMiniNode: RenderCallback<NodeType>;
  destroyCallback: DestroyCallback;
}

export default class Minimap<NodeType, EdgeType> extends React.Component< MinimapProps<NodeType, EdgeType>> {
  public render(): JSX.Element | undefined {
    const plugins = this.props.state.plugins;
    if (!plugins) return;

    const minimap = plugins.minimap;
    const workspace = this.props.state.workspace;
    const nodes = this.props.state.nodes;

    const scale = getScale(workspace, minimap);
    const containerSize = minimap.data.size;
    const canvasSize = this.getCanvasSize(workspace, containerSize, scale);
    const emptyTop = (containerSize.height - canvasSize.height) / 2;
    const emptyLeft = (containerSize.width - canvasSize.width) / 2;

    const rectOffset = getRectOffset(workspace, scale);
    const rectSize = getRectSize(workspace, scale);
    const rectTransform = `translate3d(${rectOffset.x + emptyLeft}px, ${rectOffset.y + emptyTop}px, 0)`;

    const renderedMinimapNodes = this.renderMinimapNodes(nodes, scale);

    return(
      <div
        className="dm-minimap-container"
        style={{ width: containerSize.width, height: containerSize.height }}
      >
        <div
          data-event-target={true}
          data-draggable={true}
          data-type={DiagramMakerMinimapType.RECTANGLE}
          style={{ transform: rectTransform, width: rectSize.width, height: rectSize.height }}
          className="dm-minimap-rectangle"
        />
        <div
          data-event-target={true}
          data-type={DiagramMakerMinimapType.CANVAS}
          className="dm-minimap-canvas"
          style={{ width: canvasSize.width, height: canvasSize.height, top: emptyTop, left: emptyLeft }}
        >
          {renderedMinimapNodes}
        </div>
      </div>
    );
  }

  private getCanvasSize(workspace: DiagramMakerWorkspace, containerSize: Size, scale: number): Size {
    const { height: workspaceHeight, width: workspaceWidth } = workspace.canvasSize;
    const { height: containerHeight, width: containerWidth } = containerSize;

    const workspaceRatio = workspaceWidth / workspaceHeight;
    const containerRatio = containerWidth / containerHeight;

    let canvasWidth = containerWidth;
    let canvasHeight = containerHeight;

    if (workspaceRatio > containerRatio) {
      // Short and fat, fill empty space at the top and bottom
      canvasHeight = workspaceHeight / scale;
    } else {
      // Tall and thin, fill empty space at the left and right
      canvasWidth = workspaceWidth / scale;
    }

    const canvasSize = {
      height: canvasHeight,
      width: canvasWidth
    };
    return canvasSize;
  }

  private getMinimapNode(node: DiagramMakerNode<NodeType>, scale: number): DiagramMakerNode<NodeType> {
    const { diagramMakerData: orignalDiagramMakerData, id } = node;
    const { x, y } = orignalDiagramMakerData.position;
    const { width, height } = orignalDiagramMakerData.size;
    const miniNodePos = {
      x: x / scale,
      y: y / scale
    };
    const miniNodeSize = {
      height: height / scale,
      width: width / scale
    };
    const miniNodeData = {
      position: miniNodePos,
      size: miniNodeSize
    };
    const diagramMakerData = miniNodeData;
    const miniNode = {
      diagramMakerData,
      id
    };
    return miniNode;
  }

  private renderMinimapNodes(nodes: DiagramMakerNodes<NodeType>, scale: number) {
    const renderCallback = this.props.renderMiniNode;
    const destroyCallback = this.props.destroyCallback;

    const nodeKeys = Object.keys(nodes);
    return nodeKeys.map((nodeKey: string) => {
      const miniNode = this.getMinimapNode(nodes[nodeKey], scale);
      return (
        <MiniNode
          key={`miniNode_${nodes[nodeKey].id}`}
          renderCallback={renderCallback.bind(null, nodes[nodeKey])}
          destroyCallback={destroyCallback}
          diagramMakerNode={miniNode}
        />
      );
    });
  }
}
