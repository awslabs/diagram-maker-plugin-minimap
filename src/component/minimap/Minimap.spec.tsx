import { EditorMode } from 'diagram-maker';
import Minimap from 'diagramMakerMinimap/component/minimap/Minimap';
import * as React from 'react';
import * as ShallowRenderer from 'react-test-renderer/shallow';
import './Node.scss';

describe('Minimap', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });
  const nodes = {
    node1: {
      diagramMakerData: {
        position: { x: 200, y: 150 },
        size: { width: 100, height: 50 }
      },
      id: 'node1'
    },
    node2: {
      diagramMakerData: {
        position: { x: 400, y: 300 },
        size: { width: 100, height: 50 }
      },
      id: 'node2'
    }
  };
  const minimap = {
    data:{
      size:{
        height: 200,
        width: 200
      }
    }
  };
  const plugins = { minimap };
  const edges = {};
  const panels = {};
  const editor = { mode: EditorMode.SELECT };
  const renderMininode = jest.fn();
  const destroyCallback = jest.fn();

  it('test Minimap when workspace short and wide', () => {
    const workspace = {
      canvasSize: {
        height: 1600,
        width: 3200
      },
      position: {
        x: 0,
        y: 0
      },
      scale: 1,
      viewContainerSize: {
        height: 800,
        width: 1600
      }
    };
    const state = { edges, editor, nodes, panels, plugins, workspace };

    const renderer  = ShallowRenderer.createRenderer();
    renderer.render(
      <Minimap
        state={state}
        renderMiniNode={renderMininode}
        destroyCallback={destroyCallback}
      />
    );
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  it('test Minimap when workspace tall and thin', () => {
    const workspace = {
      canvasSize: {
        height: 3200,
        width: 1600
      },
      position: {
        x: 0,
        y: 0
      },
      scale: 1,
      viewContainerSize: {
        height: 800,
        width: 1600
      }
    };
    const state = { edges, editor, nodes, panels, plugins, workspace };

    const renderer  = ShallowRenderer.createRenderer();
    renderer.render(
      <Minimap
        state={state}
        renderMiniNode={renderMininode}
        destroyCallback={destroyCallback}
      />
    );
    const result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  it('test Minimap when plugins do not exist', () => {
    const workspace = {
      canvasSize: {
        height: 3200,
        width: 1600
      },
      position: {
        x: 0,
        y: 0
      },
      scale: 1,
      viewContainerSize: {
        height: 800,
        width: 1600
      }
    };
    const state = { edges, editor, nodes, panels, workspace };

    const renderer  = ShallowRenderer.createRenderer();
    renderer.render(
      <Minimap
        state={state}
        renderMiniNode={renderMininode}
        destroyCallback={destroyCallback}
      />
    );
    const result = renderer.getRenderOutput();
    expect(result).toEqual(undefined);
  });
});
