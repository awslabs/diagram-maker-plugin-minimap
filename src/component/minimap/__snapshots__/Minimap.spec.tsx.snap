// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Minimap test Minimap when workspace short and wide 1`] = `
<div
  className="dm-minimap-container"
  style={
    Object {
      "height": 200,
      "width": 200,
    }
  }
>
  <div
    className="dm-minimap-rectangle"
    data-draggable={true}
    data-event-target={true}
    data-type="DiagramMakerMinimap.Rectangle"
    style={
      Object {
        "height": 48,
        "transform": "translate3d(0px, 50px, 0)",
        "width": 98,
      }
    }
  />
  <div
    className="dm-minimap-canvas"
    data-event-target={true}
    data-type="DiagramMakerMinimap.Canvas"
    style={
      Object {
        "height": 100,
        "left": 0,
        "top": 50,
        "width": 200,
      }
    }
  >
    <MiniNode
      destroyCallback={[MockFunction]}
      diagramMakerNode={
        Object {
          "diagramMakerData": Object {
            "position": Object {
              "x": 12.5,
              "y": 9.375,
            },
            "size": Object {
              "height": 3.125,
              "width": 6.25,
            },
          },
          "id": "node1",
        }
      }
      renderCallback={[Function]}
    />
    <MiniNode
      destroyCallback={[MockFunction]}
      diagramMakerNode={
        Object {
          "diagramMakerData": Object {
            "position": Object {
              "x": 25,
              "y": 18.75,
            },
            "size": Object {
              "height": 3.125,
              "width": 6.25,
            },
          },
          "id": "node2",
        }
      }
      renderCallback={[Function]}
    />
  </div>
</div>
`;

exports[`Minimap test Minimap when workspace tall and thin 1`] = `
<div
  className="dm-minimap-container"
  style={
    Object {
      "height": 200,
      "width": 200,
    }
  }
>
  <div
    className="dm-minimap-rectangle"
    data-draggable={true}
    data-event-target={true}
    data-type="DiagramMakerMinimap.Rectangle"
    style={
      Object {
        "height": 48,
        "transform": "translate3d(50px, 0px, 0)",
        "width": 98,
      }
    }
  />
  <div
    className="dm-minimap-canvas"
    data-event-target={true}
    data-type="DiagramMakerMinimap.Canvas"
    style={
      Object {
        "height": 200,
        "left": 50,
        "top": 0,
        "width": 100,
      }
    }
  >
    <MiniNode
      destroyCallback={[MockFunction]}
      diagramMakerNode={
        Object {
          "diagramMakerData": Object {
            "position": Object {
              "x": 12.5,
              "y": 9.375,
            },
            "size": Object {
              "height": 3.125,
              "width": 6.25,
            },
          },
          "id": "node1",
        }
      }
      renderCallback={[Function]}
    />
    <MiniNode
      destroyCallback={[MockFunction]}
      diagramMakerNode={
        Object {
          "diagramMakerData": Object {
            "position": Object {
              "x": 25,
              "y": 18.75,
            },
            "size": Object {
              "height": 3.125,
              "width": 6.25,
            },
          },
          "id": "node2",
        }
      }
      renderCallback={[Function]}
    />
  </div>
</div>
`;
